src/app/tests/TestClusterObjectUtils.cpp

```diff
@@ -0,0 +1,782 @@
+    reader.Init(mStore);
+
+    //
+    // TODO: Check-in TLV Text pretty printer in separate PR
+    //
+    //chip::TLV::Utilities::Print(reader);
+}
+
+void TestClusterObjectUtils::SetupReader()
+{
+    CHIP_ERROR err;
+
+    mReader.Init(mStore);
+    err = mReader.Next();
+
+    NL_TEST_ASSERT(mpSuite, err == CHIP_NO_ERROR);
+}
+
+template <typename T>
+struct TagValuePair {
+    uint64_t tag;
+    T& value;
+};
+
+template <typename T>
+TagValuePair<T> MakeTagValuePair(uint64_t tag, T &value) {
+    return TagValuePair<T>{tag, value};
+}
+
+template<typename... ArgTypes> CHIP_ERROR EncodeStruct(TLV::TLVWriter &writer, uint64_t tag, ArgTypes... Args)
+{
+  using expand_type = int[];
+  TLV::TLVType type;
+
+  ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Structure, type));
+  (void)(expand_type{ (ClusterObjectUtils::Encode(writer, Args.tag, Args.value), 0)... });
+  ReturnErrorOnFailure(writer.EndContainer(type));
+
+  return CHIP_NO_ERROR;
+}
+
+void TestClusterObjectUtils::TestClusterObjectUtils_EncAndDecSimpleStruct(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err;
+    chip::app::TestClusterObjectUtils *_this = static_cast<chip::app::TestClusterObjectUtils *>(apContext);
+
+    _this->mpSuite = apSuite;
+    _this->SetupBuf();
+
+    //
+    // Encode
+    //
+    {
+        chip::app::clusters::TestCluster::SimpleStruct::Type t;
+        uint8_t buf[4] = {0, 1, 2, 3};
+        char strbuf[10] = "chip";
+        
+        t.a = 20;
+        t.b = true;
+        t.c = chip::app::clusters::TestCluster::SimpleEnum::VALUEA;
+        t.d = buf;
+
+        // TODO: Bug in CHIPTLVWriter/Reader that don't quite deal with the null-terminator correctly.
+        t.e = chip::Span<char>{strbuf, strlen(strbuf)};
+        
+        err = chip::app::ClusterObjectUtils::Encode(_this->mWriter, TLV::AnonymousTag, t);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        err = _this->mWriter.Finalize();
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        _this->DumpBuf();
+    }
+
+    //
+    // Decode
+    //
+    {
+        chip::app::clusters::TestCluster::SimpleStruct::Type t;
+
+        _this->SetupReader();
+
+        err = chip::app::ClusterObjectUtils::Decode(_this->mReader, t);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        NL_TEST_ASSERT(apSuite, t.a == 20);
+        NL_TEST_ASSERT(apSuite, t.b == true);
+        NL_TEST_ASSERT(apSuite, t.c == chip::app::clusters::TestCluster::SimpleEnum::VALUEA);
+
+        for (uint32_t i = 0; i < t.d.size(); i++) {
+            NL_TEST_ASSERT(apSuite, t.d.data()[i] == i);
+        }
+
+        NL_TEST_ASSERT(apSuite, strncmp(t.e.data(), "chip", strlen("chip")) == 0);
+    }
+}
+
+void TestClusterObjectUtils::TestClusterObjectUtils_EncAndDecNestedStruct(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err;
+    chip::app::TestClusterObjectUtils *_this = static_cast<chip::app::TestClusterObjectUtils *>(apContext);
```

Comment:
This is all inside `namespace chip { namespace app {`.  Let's not have extra prefixes we don't need here.

Also, this looks like a great use case for `auto * _this`.